### PR TITLE
Move Storage to Async

### LINKED ISSUE #609: Typing issue: WebStorageStateStore.store should be an async store
I'm trying to use my own custom store implementation where all the data has to be read/write in an async manner.

```ts
userStore: new WebStorageStateStore({
  store: new MyStore()
});
```

The issue I'm facing is that [WebStorageStateStore](https://github.com/authts/oidc-client-ts/blob/main/src/WebStorageStateStore.ts) uses async methods whereas [WebStorageStateStore.store](https://github.com/authts/oidc-client-ts/blob/main/src/WebStorageStateStore.ts#L13) uses sync ones as it is of type `Storage`.

**Solutions:**
1) I could make my own implementation of `WebStorageStateStore` but then I would loose or duplicate anything that this wrapper is doing -> so I don't really like that solution.
2) It seems that supporting an async store is already implemented as all the methods from `WebStorageStateStore` returns a `Promise.resolve` of the store's results (ex: [line 34](https://github.com/authts/oidc-client-ts/blob/main/src/WebStorageStateStore.ts#L34), [line 43](https://github.com/authts/oidc-client-ts/blob/main/src/WebStorageStateStore.ts#L43), ...).
-> so the ideal solution would be to replace the type of `WebStorageStateStore.store` and make it reflect an async `Storage` type instead

What do you think ?

### PR DESCRIPTION
**Pull request by philjones88**
Closes/fixes #609

### Checklist

- [X] This PR makes changes to the public API 
- [X] I have included links for closing relevant issue numbers

### CODE REVIEW

**Comment by pamapa** [comment]
@Badisi Can you take care of reviewing this MR?

**Comment by Badisi** [comment]
@pamapa, sure! I was actually going to leave a comment ;-)

I don't know about you but I think it is a little bit too much modifications.

As I said in the original issue, the `WebStorageStateStore` can already support an async store.

If we take this for example: 
```ts
const item = this._store.getItem(key);
return Promise.resolve(item);
```

`item` could be a simple `object/value` (sync) as well as a `Promise` (async), and `Promise.resolve` will properly resolve both cases.
The only issue here is that `_store` is of type `Storage` (ie. sync). So I would just create and `AsyncStorage` type and simply change the store type to `Storage | AsyncStorage`.

----

Here is what I would do with this MR:

- [ ] keep the `InMemoryWebStorage` class (do not renaming it to avoid breaking changes for people that might already use it)
- [ ] remove `AsyncInMemoryWebStorage` and `AsyncLocalStorage` (the library should not be providing every type of storage that might exists, it should be the developer responsibility to create the one that fits his needs using the `Storage` or `AsyncStorage` types)
- [ ] revert back `OidcClientSettings`, `UserManagerSettings` and `WebStorageStateStore`
- [ ] change `WebStorageStateStore._store` type to `Storage | AsyncStorage`
- [ ] modify [`WebStorageStateStore#L25`](https://github.com/authts/oidc-client-ts/blob/1d8254343748c367ef46e0992c510759ece990a2/src/WebStorageStateStore.ts#L25) so that the result is returned by `Promise.resolve`

**Comment by philjones88** [comment]
@Badisi @pamapa I've reverted a lot of changes whilst keeping the new async storage option. Let me know if these are satisfactory.

Regarding the comment around returning `Promise.resolve` inside an `async` method, this doesn't work (Typescript complains) and isn't needed?

**Comment by Badisi** [comment]
@philjones88, looks way simpler thanks !
And is it still working like this for your use case ?

You are right it wasn't needed as the `set()` method returns `void`, my bad ^^

**Comment by philjones88** [comment]
@Badisi I believe so, it works with my scenario.

For anyone else trying to use this in an Electron React app:

For Electron I have this in my client:

```
      stateStore: new StateStoreElectron(),
      userStore: new WebStorageStateStore({ prefix: 'oidc.', store: new AsyncStoreElectron() }),
```

Where `StateStoreElectron` is:

```
import type { StateStore } from 'oidc-client-ts';
import { getLogger } from 'logger;

class StateStoreElectron implements StateStore {
  private readonly _logger = getLogger('StateStoreElectron');

  private readonly _prefix: string;

  private readonly _instance: ElectronWindow;

  public constructor({ prefix = 'oidc.' } = {}) {
    const instance = window.electron || window.parent.electron;

    if (!instance) {
      throw new Error('Window instance is undefined');
    }

    this._prefix = prefix;
    this._instance = instance;
  }

  public async set(key: string, value: string): Promise<void> {
    this._logger.debug({ key, value }, 'set');

    const keyWithPrefix = this._prefix + key;
    await this._instance.setStoreValue(keyWithPrefix, value);
    return Promise.resolve();
  }

  public async get(key: string): Promise<string | null> {
    this._logger.debug({ key }, 'get');

    const keyWithPrefix = this._prefix + key;
    const item = await this._instance.getStoreValue(keyWithPrefix);
    return Promise.resolve(item || null);
  }

  public async remove(key: string): Promise<string | null> {
    this._logger.debug({ key }, 'remove');

    const keyWithPrefix = this._prefix + key;
    const item = await this._instance.getStoreValue(keyWithPrefix);
    await this._instance.removeStoreKey(keyWithPrefix);
    return Promise.resolve(item || null);
  }

  public async getAllKeys(): Promise<string[]> {
    this._logger.debug({}, 'getAllKeys');

    const keys = await this._instance.getStoreKeys();
    return Promise.resolve(keys || []);
  }
}

export default StateStoreElectron;
```

and `AsyncStoreElectron` is:

```
/* eslint-disable no-underscore-dangle */
import { AsyncStorage } from 'oidc-client-ts';

class AsyncStoreElectron implements AsyncStorage {
  private readonly _instance: ElectronWindow;

  public get length(): Promise<number> {
    if (!window.electron) {
      return Promise.resolve(0);
    }
    return this._instance.getStoreLength();
  }

  constructor() {
    const instance = window.electron || window.parent.electron;

    if (!instance) {
      throw Error('electron on window is undefined. Are you running in Electron?');
    }

    this._instance = instance;
  }

  public async clear(): Promise<void> {
    await this._instance.clearStore();
  }

  public async getItem(key: string): Promise<string | null> {
    const item = await this._instance.getStoreValue(key);
    return item || null;
  }

  public async key(index: number): Promise<string | null> {
    const item = await this._instance.getStoreKey(index);
    return item || null;
  }

  public async removeItem(key: string): Promise<void> {
    await this._instance.removeStoreKey(key);
  }

  public async setItem(key: string, value: string): Promise<void> {
    await this._instance.setStoreValue(key, value);
  }
}

export default AsyncStoreElectron;
```

The `window.electron` and `window.parent.electron` is because inside my preload file I have:

`contextBridge.exposeInMainWorld('electron', { ... });`

This allows me to use `electron-store` to handle persistence in Electron via IPC.

Annoyingly the way OIDC-Client-TS makes iframes causes issues, so I'm having to check both the current window and parent to try and find the electron properties.

File: docs/oidc-client-ts.api.md [hunk 1]
```diff
@@ -1068,9 +1078,9 @@ export const Version: string;
 
 // @public (undocumented)
 export class WebStorageStateStore implements StateStore {
-    constructor({ prefix, store }?: {
-        prefix?: string | undefined;
-        store?: Storage | undefined;
+    constructor({ prefix, store, }?: {
```
[the remark this answers is not in the document]
Comment (philjones88) [reply docs/oidc-client-ts.api.md:1081]: eslint forces it inside `WebStorageStateStore`, if I remove it, eslint errors

````
/oidc-client-ts/src/WebStorageStateStore.ts
  19:29  error  Missing trailing comma  comma-dangle

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

[the remark this answers is not in the document]
Comment (Badisi) [reply docs/oidc-client-ts.api.md:1081]: Ok not a big deal anyway

**Review by Badisi — APPROVED** [review]
LGTM

**Review by pamapa — APPROVED** [review]

**Comment by pamapa** [comment]
thanks for the contribution and thanks for reviewing!

**Comment by andriijas** [comment]
Just wanted to double check

```
new WebStorageStateStore({ store: window.localStorage }
```

localStorage is not async interface so I need to change it with new version, or?

Thanks!

**Comment by andriijas** [comment]
Great, thanks for clarifying. Great update!

**Pull request merged**